### ISSUE TITLE
Support 'n' for Find Again

### ISSUE THREAD

**Issue by seren** [tridactyl/tridactyl#64]
Vimperator (and Vim) use lower-case `n` to jump to the next item matching a previous search. Firefox has this built-in with Command-G. Would it be possible to wire up `n` to trigger `Find Again`? Thanks.

**Comment by dumblob** [comment]
If `Command-G` somehow supports also a search in a reverse order (`N` in vim), then it sounds feasible.

**Comment by bovine3dom** [comment]
We were just going to implement our own find mode because we wanted incsearch, hlsearch and regex.

I believe @cmcaine has some code locally that he was working on.

**Comment by pholat** [comment]
I've started to use the plugin as I found it looking for Vimperator replacement 👍 
Installing the latest version of plugin I believe I saw a few times search bar and could use Vim/Vimperator like search - I'll appreciate if someone post what status is of this feature.

**Comment by bovine3dom** [comment]
@pholat we don't have it yet.

**Comment by pholat** [comment]
@bovine3dom thanks! I'm looking forward to it.

**Comment by mrlinuxfish** [comment]
I would also like this or something similar. As it is right now, I have no idea how to go to the next instance of text while searching.

**Comment by bovine3dom** [comment]
@mrlinuxfish Ctrl-G

**Comment by wschmrdrNS** [comment]
And how about in the other direction?

**Comment by bovine3dom** [comment]
@wschmrdrNS Ctrl-Shift-G

**Comment by Koushien** [comment]
NB F3 and Shift+F3 do find again, find previous respectively as well.
https://support.mozilla.org/en-US/kb/keyboard-shortcuts-perform-firefox-tasks-quickly

**Comment by wschmrdrNS** [comment]
The point of having 'n' and 'N' available is that you can continue to
search on whatever is in the find buffer after making an update to
text. You should not have to constantly re-type the same search
parameter over and over again.

**Comment by mMontu** [comment]
Any news about this issue?

**Comment by bovine3dom** [comment]
@mMontu there's a branch with most of the work done. It's blocked by the "provide feedback to users" bug: you can happily search on the branch but you have to do it blind, it doesn't tell you what you're searching for.

**Comment by mMontu** [comment]
@bovine3dom well, that is more than enough for me. Is it possible to add it to master and enable it through an option? Or the best option is to cherry pick it and do a local build?

**Comment by bovine3dom** [comment]
Potentially as an option. I would need to think about "/" as that is currently a special key that gets sent to Firefox. It would happily work for something nearby, e.g. "." though.

**Comment by dumblob** [comment]
>It would happily work for something nearby, e.g. "." though.

I have to admit I don't like this idea of using (any) other key instead of `/`. What about supporting just `?` for the time being (before Firefox allows redefining `/`) and telling users to use `N` instead of `n`?

**Comment by bovine3dom** [comment]
@mMontu Tridactyl versions >=1.8.2-25-g682719c (i.e, the latest beta releases) now have the `find` and `findnext` excmds.

Because I didn't think about it for very long, `find` shoves you in a find mode. It does not accept a query as an argument, but the direction. e.g:
`bind . find`
`bind ? find -1`
`bind n findnext`
`bind N findnext -1`.

Is I explained above, binding to `/` is not currently possible as it would stop the Firefox quick find from working.

Other known issues:
- performance is not ideal
- highlighted option always takes you to top of page rather than starting from elements in viewport

---

@dumblob I think you misunderstood me. We will bind it to `/` by default once it is good enough for most people. We send `/` to Firefox on purpose because we think their quick find bar is better than ours.

**Comment by wschmrdrNS** [comment]
One other hurdle I found, and it just started today, looks more to do with the behavior of the browser itself than Tridactyl but would be nice to get past if it's possible:

On some pages (such as in Github; not sure if you can actually test it on this one or not), I noticed that the search command, whether by using '/' or the shortcut suggested in the comments previous, sends the user to the search box on the page itself, rather than the browser or Tridactyl search. Presently 1.8.2.25-g682719c, Firefox 58.0.2 64-bit.

**Comment by mMontu** [comment]
> On some pages (such as in Github; not sure if you can actually test it on this one or not), I noticed that the search command, whether by using '/' or the shortcut suggested in the comments previous, sends the user to the search box on the page itself

@wschmrdrNS I started seeing this today -- probably triggered by some recent update on GitHub, as I'm pretty sure this wasn't happening yesterday.

**Comment by bovine3dom** [comment]
I typed this up yesterday but apparently forgot to post it...

---

> the n sometimes fail, as if it have find something in the page, but failed to move the page or update the highlight (my guess is it highlighted something in the source that isn't marked as visible). It happens when searching in both directions

Yeah, I noticed this - it's particularly bad on GitHub.

> searching backward (?) sometimes makes the next forward search to fail -- .find<cr>n takes me to the second occurrence, but the next n jumps to the last one (which currently is in the last sentence of the previous comment). Then it starts to go backwards, as if I was typing N instead of n

I can't reproduce this - which site were you testing it on?

**Comment by mMontu** [comment]
> I can't reproduce this - which site were you testing it on?

I tested this current page, I mean, this issue (#64) on GitHub.

**Comment by bovine3dom** [comment]
@mMontu 1.8.2-29 should be more useful on GitHub now - you can only select visible text.

It's still quite slow because it searches through lots of text that isn't visible, and I can't figure out how to ignore "undrawn" text (if anyone wants a go, please feel free - my attempt is in 1441682eff9a150e2a703fad318a775e0bc64dac and I have left some unhelpful comments).

The "/" on GitHub isn't something we can fix until we replace find mode with our own, as far as I can tell.

**Comment by mMontu** [comment]
thanks @bovine3dom, I'll give it a try

**Comment by gshrikant** [comment]
There is probably a workaround for this but with the current '/' execmd (which opens Firefox quick find) I almost always end up focusing on the link itself and pressing Enter followed by 'n' or 'N' (bound to functions in the previous post) show `Error processing navigate() TypeError: findModeState is undefined`.

**Comment by bovine3dom** [comment]
@gshrikant Use Ctrl-G / Ctrl-Shift-G instead : )

**Comment by GianniGi** [comment]
Still hoping that n/N will be implemented. Ctrl-g is annoying...

**Comment by sandangel** [comment]
any update on this? please add n/N, as my vim memory muscle has get used to it

**Comment by bovine3dom** [comment]
We were just talking about it yesterday - https://github.com/tridactyl/tridactyl/issues/1000#issuecomment-425620535.

It's quite a long way off as it's blocked by quite a few other issues.

**Comment by jjzz** [comment]
I tried to use xdotool to send 'F3' key (Find Next) to firefox (firefox-esr v60 @ debian)

    bind n exclaim_quiet xdotool getactivewindow key F3           # fail
    bind N exclaim_quiet xdotool getactivewindow key Shift+F3     # work

Not sure why 'N' worked but 'n' failed.

A further test on 'j' and 'n' with a simpler task,

    bind n scrollline 8     # fail
    bind j scrollline 8     # work

It seems something wrong with the binding on key 'n'?

**Comment by bovine3dom** [comment]
`bind n scrollline 8` works fine for me. Could you open another issue and we can look into it further?

Using `xdotool` is a very good idea, by the way.

**Comment by bovine3dom** [comment]
Closing this as latest Betas have an OK find mode. Still no incsearch but I'll leave #1612 open for that.
